src/kv/kv_types.h

```diff
@@ -93,13 +93,55 @@ namespace kv
     Nodes nodes;
   };
 
+  inline void to_json(nlohmann::json& j, const Configuration::NodeInfo& ni)
+  {
+    j["address"] = fmt::format("{}:{}", ni.hostname, ni.port);
+  }
+
+  inline void from_json(const nlohmann::json& j, Configuration::NodeInfo& ni)
+  {
+    const std::string addr(j["address"]);
+    const auto& [h, p] = nonstd::split_1(addr, ":");
+    ni.hostname = h;
+    ni.port = p;
+  }
+
+  enum class ReplicaState
+  {
+    Leader,
+    Follower,
+    Candidate,
+    Retired
+  };
+
+  DECLARE_JSON_ENUM(
+    ReplicaState,
+    {{ReplicaState::Leader, "Leader"},
+     {ReplicaState::Follower, "Follower"},
+     {ReplicaState::Candidate, "Candidate"},
+     {ReplicaState::Retired, "Retired"}});
+
+  DECLARE_JSON_TYPE(Configuration);
+  DECLARE_JSON_REQUIRED_FIELDS(Configuration, idx, nodes);
+
+  struct ConsensusDetails
+  {
+    std::vector<Configuration> configs = {};
+    std::unordered_map<ccf::NodeId, ccf::SeqNo> acks = {};
```

Comment:
Is the idea that only trusted nodes are listed in the configs? For me it would be convenient to add a node state field to indicate whether they are pending/joining/exiting/etc.

Reply:
Yes, only trusted nodes in a given config are listed here, because that's all consensus ever deals with. Nodes that aren't trusted yet definitely do not have the ability to verify or decrypt append entries, so talking to them isn't very useful.

Nodes that are retiring are in a config, but not in later configs, so they show up here up until the point where their removal is committed and the config dropped.